compiler/qsc_qasm3/src/parse.rs

```diff
@@ -0,0 +1,265 @@
+/// and parse results.
+fn create_source_map(source: &QasmSource) -> SourceMap {
+    let mut files: Vec<(Arc<str>, Arc<str>)> = Vec::new();
+    files.push((
+        Arc::from(source.path().to_string_lossy().to_string()),
+        Arc::from(source.source()),
+    ));
+    // Collect all source files from the includes, this
+    // begins the recursive process of collecting all source files.
+    for include in source.includes() {
+        collect_source_files(include, &mut files);
+    }
+    // Map the main source file to the entry point expression
+    // This may be incorrect, but it's the best we can do for now.
+    SourceMap::new(files, Some(Arc::from(source.source())))
+}
+
+/// Recursively collect all source files from the includes
+fn collect_source_files(source: &QasmSource, files: &mut Vec<(Arc<str>, Arc<str>)>) {
+    files.push((
+        Arc::from(source.path().to_string_lossy().to_string()),
+        Arc::from(source.source()),
+    ));
+    for include in source.includes() {
+        collect_source_files(include, files);
+    }
+}
+
+/// Represents a QASM source file that has been parsed.
+#[derive(Clone, Debug)]
+pub struct QasmSource {
+    /// The path to the source file. This is used for error reporting.
+    /// This path is just a name, it does not have to exist on disk.
+    path: PathBuf,
+    /// The source code of the file.
+    source: String,
+    /// The parsed AST of the source file or any parse errors.
+    ast: ParseOrErrors<oq3_syntax::SourceFile>,
+    /// Any included files that were resolved.
+    /// Note that this is a recursive structure.
+    included: Vec<QasmSource>,
+}
+
+pub trait QasmSourceTrait {
+    fn has_errors(&self) -> bool {
+        if !self.parse_result().errors().is_empty() {
+            return true;
+        }
+        self.includes().iter().any(QasmSourceTrait::has_errors)
+    }
+    fn all_errors(&self) -> Vec<crate::Error> {
+        let mut self_errors = self.errors();
+        let include_errors = self.includes().iter().flat_map(QasmSourceTrait::all_errors);
+        self_errors.extend(include_errors);
+        self_errors
+    }
+    fn includes(&self) -> &Vec<QasmSource>;
+    fn source(&self) -> &str;
+    fn parse_result(&self) -> &ParseOrErrors<oq3_syntax::SourceFile>;
+    fn errors(&self) -> Vec<crate::Error>;
+    fn path(&self) -> PathBuf;
+
+    fn tree(&self) -> oq3_syntax::SourceFile {
+        self.parse_result().tree()
+    }
+    fn syntax_node(&self) -> SyntaxNode {
+        self.parse_result().syntax_node()
+    }
+}
+
+impl QasmSource {
+    pub fn new<T: AsRef<str>, P: AsRef<Path>>(
+        source: T,
+        file_path: P,
+        ast: ParseOrErrors<oq3_syntax::SourceFile>,
+        included: Vec<QasmSource>,
+    ) -> QasmSource {
+        QasmSource {
+            source: source.as_ref().to_owned(),
+            path: file_path.as_ref().to_owned(),
+            ast,
+            included,
+        }
+    }
+}
+
+impl QasmSourceTrait for QasmSource {
```

Comment:
Based on my cmd+F-ing, this is only implemented for one type. Could this trait just be methods on `QasmSource`?